fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/metadata/RecordMetaDataBuilderTest.java

```diff
@@ -370,4 +420,213 @@ public void updateRecords() {
+        assertEquals(4L, metaData.getIndex("MySimpleRecord$num_value_2").getSubspaceKey());
+
+        // Valid use of counter-based subspace keys assignment with meta-data proto.
+        builder = RecordMetaData.newBuilder().enableCounterBasedSubspaceKeys().setRecords(metaData.toProto());
+        metaData = builder.build(true);
+        assertNotNull(metaData.getIndex("MySimpleRecord$str_value_indexed"));
+        assertNotNull(metaData.getIndex("MySimpleRecord$num_value_unique"));
+        assertNotNull(metaData.getIndex("MySimpleRecord$num_value_3_indexed"));
+        assertNotNull(metaData.getIndex("MySimpleRecord$num_value_2"));
+        assertEquals(1L, metaData.getIndex("MySimpleRecord$str_value_indexed").getSubspaceKey());
+        assertEquals(2L, metaData.getIndex("MySimpleRecord$num_value_unique").getSubspaceKey());
+        assertEquals(3L, metaData.getIndex("MySimpleRecord$num_value_3_indexed").getSubspaceKey());
+        assertEquals(4L, metaData.getIndex("MySimpleRecord$num_value_2").getSubspaceKey());
+
+        // FormerIndex
+        RecordMetaDataBuilder formerIndexBuilder = RecordMetaData.newBuilder().enableCounterBasedSubspaceKeys().setRecords(TestRecords1Proto.getDescriptor());
+        Object formerSubspaceKey = formerIndexBuilder.getIndex("MySimpleRecord$num_value_3_indexed").getSubspaceKey();
+        formerIndexBuilder.removeIndex("MySimpleRecord$num_value_3_indexed");
+        final RecordMetaData metaDataWithFormerIndex = formerIndexBuilder.build(true);
+        FormerIndex formerIndex = metaDataWithFormerIndex.getFormerIndexes().stream().filter(index -> index.getFormerName().equals("MySimpleRecord$num_value_3_indexed")).findFirst().get();
+        assertEquals(formerSubspaceKey, formerIndex.getSubspaceKey());
+
+        // A common case that user had some existing meta-data in which all indexes had implicit subspace keys, and the user now decides to enable this feature.
+        RecordMetaDataProto.MetaData.Builder protoBuilder = RecordMetaDataProto.MetaData.newBuilder()
+                .setRecords(TestRecords1Proto.getDescriptor().toProto());
+        for (RecordMetaDataProto.Index.Builder index : protoBuilder.getIndexesBuilderList()) {
+            index.clearSubspaceKey();
+        }
+        builder = RecordMetaData.newBuilder().enableCounterBasedSubspaceKeys().setRecords(protoBuilder.build(), true);
+        metaData = builder.build(true);
+        assertNotNull(metaData.getIndex("MySimpleRecord$str_value_indexed"));
+        assertNotNull(metaData.getIndex("MySimpleRecord$num_value_unique"));
+        assertNotNull(metaData.getIndex("MySimpleRecord$num_value_3_indexed"));
+        assertEquals(1L, metaData.getIndex("MySimpleRecord$str_value_indexed").getSubspaceKey());
+        assertEquals(2L, metaData.getIndex("MySimpleRecord$num_value_unique").getSubspaceKey());
+        assertEquals(3L, metaData.getIndex("MySimpleRecord$num_value_3_indexed").getSubspaceKey());
```

Comment:
Wait, shouldn't these be set to the names, not the counter values?

Reply:
no, they have no subspace keys. the counter-based scheme is enabled, so all indexes will be assigned counters.

Reply:
Oh, I see. This isn't actually the test I wanted; that's the problem.

In this test, there aren't any indexes with no subspace keys; `protoBuilder.getIndexesBuilderList()` is empty, so all of the indexes are from processing the extension options (which are new indexes from after the option is enabled).

What I'm looking for is existing indexes in the meta-data proto which don't have subspace keys. Something like:

```java
RecordMetaDataProto.MetaData.Builder protoBuilder = metaData.toProto().toBuilder();
protoBuilder.getIndexesBuilderList().forEach(RecordMetaDataProto.Index.Builder::clearSubspaceKey);
metaData = ReocrdMetaData.newBuilder().enableCounterBasedSubspaceKeys().setRecords(protoBuilder.build()).getRecordMetaData();
```

In that case, the indexes should have their names as subspace keys.

Reply:
sure, done.